### PR TITLE
Sort Arrays by keys before generating callable checksums

### PR DESCRIPTION
**Pull request by mdbitz**
The checksum we generate are different based on the order of Associated Arrays. This update applies a ksort on arrays to ensure we have consistent order.

Fixes # https://wp.me/p1HpG7-9UI

#### Changes proposed in this Pull Request:

* Applies a ksort when generating checksums to ensure duplicate updates are not sent for callables.

#### Does this pull request change what data or activity we track or use?

NO

#### Testing instructions:

* Checkout this branch
* Perform an update of a post (triggers sync)
* Verify that no callables where sent in Sync (Activity Log)
* Update a plugin (perform a change that modifies get_plugins callable function)
* Verify that the callable is sent via Sync
* Verify that in Activity Log we see a plugin__updated entry for the plugin

Perform updates to other callables (delete plugin, home/site_url, etc) and confirm they are sent via Sync.

#### Proposed changelog entry for your changes:

* Sync Bug Fix : Remove duplicate Activity Log entries by applying ksort during checksum generation.

#### Additional Notes

Please be aware that this change impacts how we determine if data should be sent to WP.com. It needs additional testing/scrutiny to ensure that no duplicate activity is generated for a site. 

Also from a development perspective is ksort the most effective approach. @bisko suggested the below which would be faster although I'm not sure if it would lead to additional or less collisions with the checksum.

```
$data = serialize( $values );
$data = explode('', $data);
$data = sort($data);
$checksum = md5($data);
```

### CODE REVIEW

**Comment by mdbitz** [comment]
Crew Note :: The --no-verify flag was used as the Module class has several (WordPress.DB.PreparedSQL.InterpolatedNotPrepared) phpcs issues flagged which are existing code and not part of this PR.

File: packages/sync/src/modules/class-module.php [hunk 1]
```diff
@@ -179,12 +179,31 @@ protected function count_actions( $action_names, $actions_to_count ) {
 	 * @access protected
 	 *
 	 * @param mixed $values Values to calculate checksum for.
+	 * @param bool  $sort If $values should have ksort called on it.
```
Comment (bisko) [inline packages/sync/src/modules/class-module.php:182]: Do we need to have this parameter? Do you see a place where we don't want to perform the sort, except to keep up with legacy sites that do not use the sorting?

Even if we force-sort the values, the worst case scenario would be that we will be syncing the callables for all those sites that upgrade and have issues with the sorted arrays. This should be a one-time deal and will keep the code complexity down.

**Review by bisko — DISMISSED** [review]

File: packages/sync/src/modules/class-module.php (hunk 1 above)
Comment (mdbitz) [reply packages/sync/src/modules/class-module.php:182]: We really don't want there to be additional activity in the Activity Log or other downstream systems. To get around this we could potentially tie into the upgrade routine and have it go through and update all the checksums. However as that would be in the update routine itself we would still want to ensure that legit changes did get sent across.

File: packages/sync/src/modules/class-module.php [hunk 2]
```diff
@@ -179,12 +179,31 @@ protected function count_actions( $action_names, $actions_to_count ) {
...
+		}
 		return crc32( wp_json_encode( jetpack_json_wrap( $values ) ) );
 	}
 
+	/**
+	 * Recursively call ksort on an Array
+	 *
+	 * @param array $values Array.
+	 */
+	private function recursive_ksort( $values ) {
```
[the remark this answers is not in the document]
Comment (mdbitz) [reply packages/sync/src/modules/class-module.php]: Resolved in latest commit, this is why I should push unit tests before marking ready for review.

**Comment by mdbitz** [comment]
Ready again for thoughts/feedback :)

File: tests/php/sync/test_class.jetpack-sync-callables.php [hunk 3]
```diff
@@ -1151,6 +1151,49 @@ public function test_get_main_network_site_wpcom_id_multisite() {
...
+	 *
+	 * @return array
+	 */
+	public function randomize_array_keys( $plugins ) {
+		// First plugin in array.
+		$plugin_key = array_keys( $plugins )[0];
+
+		// sort the plugin's array entries.
+		$keys = array_keys( $plugins[ $plugin_key ] );
+		shuffle( $keys );
```
Comment (anomiex) [inline tests/php/sync/test_class.jetpack-sync-callables.php:1188]: It's possible that this doesn't actually change the order of the keys, particularly if `$keys` has few entries.

In general, randomness in tests is bad as it makes test runs non-reproducible.

**Review by anomiex — COMMENTED** [review]
>  @bisko suggested the below which would be faster although I'm not sure if it would lead to additional or less collisions with the checksum.
> ```
> $data = serialize( $values );
> $data = explode('', $data);
> $data = sort($data);
> $checksum = md5($data);
> ```

Seems like a risky idea to me. It would produce the same checksum if a value changes from "12" to "21", or two keys swap values, etc.

File: tests/php/sync/test_class.jetpack-sync-callables.php (hunk 3 above)
Comment (mdbitz) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: Excellent point, updated to use array_reverse and renamed filter function to align to no longer being random.

Thanks

**Comment by mdbitz** [comment]
Ready again for Crew Review :)

File: tests/php/sync/test_class.jetpack-sync-callables.php [hunk 4]
```diff
@@ -1216,25 +1216,23 @@ public function test_sync_does_not_send_updates_if_array_order_is_only_change()
...
-		$random = array();
-		foreach ( $keys as $key ) {
-			$random[ $key ] = $plugins[ $plugin_key ][ $key ];
-		}
-		$plugins[ $plugin_key ] = $random;
-		return $plugins;
+		// reverse the plugin's array entries.
+		$keys                   = array_keys( $plugins[ $plugin_key ] );
+		$keys                   = array_reverse( $keys );
+		$plugins[ $plugin_key ] = $keys;
```
Comment (anomiex) [inline tests/php/sync/test_class.jetpack-sync-callables.php:1232]: This seems to not be doing the right thing. Just
```suggestion
		$plugins[ $plugin_key ] = array_reverse( $plugins[ $plugin_key ] );
```
should be sufficient.

File: tests/php/sync/test_class.jetpack-sync-callables.php [hunk 5]
```diff
@@ -1216,25 +1216,23 @@ public function test_sync_does_not_send_updates_if_array_order_is_only_change()
...
 	/**
-	 * Randomize the get_plugins array keys.
+	 * Reorder the get_plugins array keys.
 	 *
 	 * @param array $plugins array of plugins.
 	 *
 	 * @return array
 	 */
-	public function randomize_array_keys( $plugins ) {
+	public function reverse_array_keys( $plugins ) {
```
Comment (anomiex) [inline tests/php/sync/test_class.jetpack-sync-callables.php:1225]: You called it "reorder_array_keys" above, but "reverse_array_keys" here.
```suggestion
	public function reorder_array_keys( $plugins ) {
```

**Review by anomiex — CHANGES_REQUESTED** [review]

File: tests/php/sync/test_class.jetpack-sync-callables.php (hunk 4 above)
Comment (mdbitz) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: 🎩 I know better than to quickly make changes at random hours. Thanks for the thorough look over.

**Review by anomiex — APPROVED** [review]
Seems ok to me now

**Review by kraftbj — APPROVED** [review]

**Pull request merged**